junit/src/main/java/cucumber/runtime/junit/PickleRunners.java

```diff
@@ -0,0 +1,227 @@
+package cucumber.runtime.junit;
+
+import cucumber.runner.Runner;
+import gherkin.events.PickleEvent;
+import gherkin.pickles.PickleStep;
+import org.junit.runner.Description;
+import org.junit.runner.notification.RunNotifier;
+import org.junit.runners.ParentRunner;
+import org.junit.runners.model.InitializationError;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+
+class PickleRunners {
+
+    interface PickleRunner {
+        void run(RunNotifier notifier);
+
+        Description getDescription();
+
+        Description describeChild(PickleStep step);
+
+    }
+
+    static PickleRunner withStepDescriptions(cucumber.runner.Runner runner, PickleEvent pickleEvent, JUnitReporter jUnitReporter) throws InitializationError {
+        return new WithStepDescriptions(runner, pickleEvent, jUnitReporter);
+    }
+
+
+    static PickleRunner withNoStepDescriptions(cucumber.runner.Runner runner, PickleEvent pickleEvent, JUnitReporter jUnitReporter) throws InitializationError {
+        return new NoStepDescriptions(runner, pickleEvent, jUnitReporter);
+    }
+
+
+    static class WithStepDescriptions extends ParentRunner<PickleStep> implements PickleRunner {
+        private final Runner runner;
```

Comment:
The reason for two separate classes is because the semantics of ParentRunner somewhat expect there to be children. It can be hacked but doing so would make things fragile.